File: notification_service_js/dependency-injection-factory/NotificationService.js
class NotificationService {
  constructor(fileHandler) {
    this.fileHandler = fileHandler;
  }

  sendEmail(recipient, message) {
    const notification = {
      type: 'email',
      recipient,
      message,
      timestamp: new Date(),
    };
    this.saveNotification(notification);
    console.log(`Email sent to ${recipient}: ${message}`);
  }

  sendSMS(recipient, message) {
    const notification = {
      type: 'sms',
      recipient,
      message,
      timestamp: new Date(),
    };
    this.saveNotification(notification);
    console.log(`SMS sent to ${recipient}: ${message}`);
  }

  sendPushNotification(recipient, message) {
    const notification = {
      type: 'push',
      recipient,
      message,
      timestamp: new Date(),
    };
    this.saveNotification(notification);
    console.log(`Push notification sent to ${recipient}: ${message}`);
  }

  saveNotification(notification) {
    const notifications = this.fileHandler.readFile();
    notifications.push(notification);
    this.fileHandler.writeFile(notifications);
  }

  getNotifications() {
    return this.fileHandler.readFile();
  }
}

module.exports = NotificationService;

File: notification_service_rb/example/notification_service.rb
require_relative 'file_handler'

class NotificationService
  def initialize(file_path)
    @file_handler = FileHandler.new(file_path)
  end

  def send_email(recipient, message)
    notification = {
      type: 'email',
      recipient: recipient,
      message: message,
      timestamp: Time.now
    }
    save_notification(notification)
    puts "Email sent to #{recipient}: #{message}"
  end

  def send_sms(recipient, message)
    notification = {
      type: 'sms',
      recipient: recipient,
      message: message,
      timestamp: Time.now
    }
    save_notification(notification)
    puts "SMS sent to #{recipient}: #{message}"
  end

  def send_push_notification(recipient, message)
    notification = {
      type: 'push',
      recipient: recipient,
      message: message,
      timestamp: Time.now
    }
    save_notification(notification)
    puts "Push notification sent to #{recipient}: #{message}"
  end

  def get_notifications
    @file_handler.read_file
  end

  private

  def save_notification(notification)
    notifications = @file_handler.read_file
    notifications << notification
    @file_handler.write_file(notifications)
  end
end
